README.md
[![Build Status](https://travis-ci.org/mskilab/gUtils.svg?branch=master)](https://travis-ci.org/mskilab/gUtils)
[![Documentation Status](https://readthedocs.org/projects/gutils/badge/?version=latest)](https://readthedocs.org/projects/gutils/?badge=latest)
[![Coverage Status](https://coveralls.io/repos/github/mskilab/gUtils/badge.svg?branch=master)](https://coveralls.io/github/mskilab/gUtils?branch=master)


gUtils
=======

Set of utility functions for use with GenomicRanges



Installation
------------

1. Install dependent packages and latest Bioconductor (if you haven't already)
  ```source("http://bioconductor.org/biocLite.R") 
     bioclite()
     biocLite(c("GenomicRanges"))
  ```

2. Install devtools from CRAN (if you don't have it already)

  ```
  install.packages('devtools')
  ```

3. Load devtools

  ```
  library(devtools)
  ````

4. Install gUtils

  ```
  install_github('mskilab/gUtils')
  ````

Quickstart
----------

Usage
-----

Full documentation with examples is available here: [Documentation][docs]

Attributions
------------
> Marcin Imielinski - Assistant Professor, Weill-Cornell Medical College. Developed while a fellow in the Matthew Meyerson Lab, Dana Farber Cancer Institute, Broad Institute

> Jeremiah Wala - Harvard MD-PhD candidate, Bioinformatics and Integrative Genomics, Rameen Beroukhim Lab, Dana Farber Cancer Institute

[license]: https://github.com/jwalabroad/gTrack/blob/master/LICENSE

[docs]: http://gutils.readthedocs.org/en/latest/index.html



NEWS.md
gUtils 0.1.0 2016-03-18
-----------------------------------------------------------

* Initial stable release to Github


tests/testthat.R
install.packages(testthat)
library(testthat)
library(gUtils)

test_check("gUtils")


tests/testthat/test_rangeops.R
library(gUtils)
library(BSgenome.Hsapiens.UCSC.hg19)
context("Range ops")

gr  <- GRanges(1, IRanges(c(3,7,13), c(5,9,16)), strand=c('+','-','-'), seqinfo=Seqinfo("1", 25), name=c("A","B","C"))
gr2 <- GRanges(1, IRanges(c(1,9), c(6,14)), strand=c('+','-'), seqinfo=Seqinfo("1", 25), field=c(1,2))
dt <- data.table(seqnames=1, start=c(2,5,10), end=c(3,8,15))

test_that("hg_seqlengths", {
  expect_error(hg_seqlengths())
  ee <- structure(names="1", 249250621L)
  expect_identical(hg_seqlengths(Hsapiens)[1],ee)
  expect_equal(names(hg_seqlengths(Hsapiens, chr=TRUE)[1]), "chr1")
  expect_equal(length(hg_seqlengths(Hsapiens, include.junk = TRUE)), 93)
})

test_that("gr.start ", {
  expect_identical(start(gr.start(gr)), c(3L,7L,13L))
  expect_identical(end(gr.start(gr)),   c(3L,7L,13L))
  expect_identical(end(gr.start(gr, width=100)),   c(25L,25L,25L))
  expect_identical(suppressWarnings(end(gr.start(gr, width=100, force=TRUE))),   c(102L,106L,112L))
  expect_identical(end(gr.start(gr, width=100, ignore.strand=FALSE)),   c(25L,9L,16L))
  expect_identical(suppressWarnings(start(gr.start(gr, width=100, ignore.strand=FALSE, force=TRUE))),   c(3L,-90L,-83L))
  expect_identical(suppressWarnings(end(gr.start(gr, width=100, force=TRUE))),   c(102L,106L,112L))

  expect_identical(end(gr.start(gr, width=10000, clip=TRUE)), c(5L, 9L, 16L))
})

test_that("gr.end", {

  expect_identical(start(gr.end(gr)), c(5L,9L,16L))
  expect_identical(start(gr.end(gr, width=100)),   c(1L,1L,1L))
  expect_identical(suppressWarnings(start(gr.end(gr, width=100, force=TRUE))),   c(-94L,-90L,-83L))
  expect_identical(start(gr.end(gr, width=100, ignore.strand=FALSE)),   c(1L,7L,13L))
  expect_identical(suppressWarnings(start(gr.end(gr, width=100, ignore.strand=FALSE, force=TRUE))),   c(-94L,7L,13L))

})

test_that("gr.mid", {
  expect_identical(start(gr.mid(gr)), c(4L,8L,14L))
})

test_that("gr.dist", {
  m <- gr.dist(gr, gr2, ignore.strand=TRUE)
  expect_equal(m[1,2], 3)
})

test_that("gr.rand", {
  set.seed(137)
  gg <- gr.rand(c(3,5), si)
  print(start(gg))
  expect_equal(start(gg)[1], 59221325L)
})

test_that("si2gr", {
  gg <- si2gr(si, strip.empty = TRUE)
  expect_equal(start(gg)[3], 1)
  expect_equal(end(gg)[1], 249250621)
  expect_equal(as.character(strand(gg)[1]), "+")
})

test_that("grbind", {
  expect_that(length(grbind(example_genes, example_dnase)) > 0, is_true())
  expect_equal(colnames(mcols(grbind(example_genes, example_dnase)))[1], "name")
})

test_that("gr.dice", {
  expect_equal(length(gr.dice(gr)[[3]]),4)
})

test_that("gr.findoverlaps", {
  fo <- gr.findoverlaps(example_genes, example_dnase)
  expect_equal(ncol(mcols(fo)), 2)
  expect_that(length(fo) > 0, is_true())

  ## null input
  expect_equal(length(gr.findoverlaps(example_genes, GRanges())), 0)
  expect_equal(length(gr.findoverlaps(GRanges(), GRanges())), 0)
  expect_equal(length(gr.findoverlaps(GRanges(), example_dnase)), 0)
})

test_that("gr.findoverlaps, return as data.table", {
  expect_error(gr.findoverlaps(example_genes, example_dnase, return.type = "data.frame"))

  fo <- gr.findoverlaps(example_genes, example_dnase, return.type = 'data.table')
  expect_identical(colnames(fo), c("start", "end", "query.id", "subject.id", "seqnames", "strand"))
})

test_that("rrbind", {
  expect_that(ncol(rrbind(mcols(example_genes), mcols(example_dnase))) > 2, is_true())
  expect_equal(ncol(rrbind(mcols(example_genes), mcols(example_dnase), union=FALSE)), 0)
})

test_that("gr.match", {
  ## accepts data.table
  expect_that(sum(!is.na(gr.match(gr2dt(example_genes), example_genes))) > 0, is_true())

  ## gives back overlapping matches
  gr11 <- GRanges(1, IRanges(c(10,20), width=5), strand=c("+", "-"))
  gr12 <- GRanges(1, IRanges(c(8,18, 100), width=5), strand=c("-", "+", "+"))

  expect_identical(gr.match(gr11, gr12), c(1L,2L))

  ## ignore strand is successfully passed
  expect_identical(gr.match(gr11, gr12, ignore.strand = FALSE), c(NA,NA))
})

test_that("gr.findoverlaps chunk", {
  fo  <- gr.findoverlaps(example_genes, example_dnase)
  fo2 <- gr.findoverlaps(example_genes, example_dnase, max.chunk = 1e7, verbose=TRUE)
  expect_identical(fo, fo2)
})

test_that("gr.findoverlaps, input data.table", {
  expect_equal(class(gr.findoverlaps(gr2dt(example_genes), example_genes, return.type='GRanges'))[1], "GRanges")
  expect_equal(class(gr.findoverlaps(gr2dt(example_genes), example_genes))[1], "data.table")
  expect_equal(class(gr.findoverlaps(gr2dt(example_genes), example_genes, max.chunk = 1e7))[1], "data.table")
})

test_that("gr.findoverlaps ignore.strand", {

  ## make a stranded DNAase track (for testing only)
  example_dnase2 = example_dnase
  set.seed(137)
  strand(example_dnase2) <- ifelse(runif(length(example_dnase)) > 0.5, '+', '-')

  ## get the overlaps with the original unstranded, and with ignore.strand
  fo1 <- gr.findoverlaps(example_dnase, example_genes)
  fo2 <- gr.findoverlaps(example_dnase2, example_genes, ignore.strand=TRUE)
  expect_identical(fo1, fo2)

  ## make sure no strands overlap
  fo1 <- gr.findoverlaps(example_dnase2, example_genes, ignore.strand=FALSE)
  expect_that(!any(strand(example_dnase2)[fo1$query.id] != strand(example_genes)[fo1$subject.id]), is_true())
})

# test_that("gr.findoverlaps foverlaps", {
#   ## make stranded DNAase track (for testing only)
#   example_dnase2 = example_dnase
#   set.seed(137)
#   strand(example_dnase2) <- ifelse(runif(length(example_dnase)) > 0.5, '+', '-')
#
#   ## assure that things are same with/without foverlaps
#   fo1 <- gr.findoverlaps(example_dnase, example_genes, foverlaps=TRUE)
#   fo1b <- gr.findoverlaps(example_genes, example_dnase, foverlaps=TRUE)
#   expect_identical(fo1$query.id, fo1b$subject.id)
#   expect_identical(fo1$subject.id, fo1b$query.id)
#   expect_identical(start(fo1), start(fo1b))
#
#   fo2 <- gr.findoverlaps(example_dnase, example_genes, foverlaps=FALSE)
#   fo2b <- gr.findoverlaps(example_genes, example_dnase, foverlaps=FALSE)
#   expect_identical(fo1$query.id, fo1b$subject.id)
#   expect_identical(fo1$subject.id, fo1b$query.id)
#   expect_identical(start(fo1), start(fo1b))
#
#   expect_identical(start(fo1), start(fo2))
#   expect_identical(end(fo1), end(fo2))
#   ## sort subject id because order for multi-overlaps is different, but it's OK
#   expect_identical(sort(fo1$subject.id), sort(fo2$subject.id))
# })

test_that("gr.findoverlap by", {
  e1 <- example_genes
  e2 <- example_dnase
  e1$bin <- e2$bin <- 1
  expect_error(gr.findoverlaps(example_genes, example_dnase, by = "dummy"))
  expect_that(length(gr.findoverlaps(e1, e2, by = "bin")) > 0, is_true())
})

test_that("gr2dt works as expected", {
  expect_identical(colnames(gr2dt(gr)), c("seqnames","start",'end','width','strand','name'))
  expect_equal(nrow(gr2dt(gr)), length(gr))

  #subjectdt <- gr2dt(example_genes)
  #expect_that(!any(subjectdt$start!=start(example_genes)), is_true)
  #expect_that(!any(subjectdt$end!=end(example_genes)), is_true)
  #expect_that(!any(subjectdt$width!=width(example_genes)), is_true)
  #expect_that(!any(subjectdt$strand!=strand(example_genes)), is_true)
  #expect_that(!any(subjectdt$seqnames!=seqnames(example_genes)), is_true)
})

test_that("dt2gr", {
  dt <- data.table(seqnames=1, start=1, end=10, strand='+', name="A")
  expect_equal(as.character(strand(dt2gr(dt))), '+')
  expect_equal(start(dt2gr(dt)), 1)
  expect_equal(dt2gr(dt)$name, "A")

  dt <- data.table(seqns=1, start=1, end=10, strand='+', name="A")
  expect_error(dt2gr(dt))

})

test_that("gr.sample", {
  set.seed(137)
  gg <- gr.sample(reduce(example_genes), 10, len=1)
  expect_equal(unique(width(gg)), 1)

  ## query width less than output
  expect_error(gr.sample(gr.start(example_genes), c(1:3), len=5)) 
  
  gg <- gr.sample(example_genes[1:5], c(2,2,3,4,5), len=2)
  expect_equal(length(gg), 16)
  expect_equal(sum(width(gg)), 32)

})


test_that("gr.sample without replace", {
  gg <- gr.sample(reduce(example_genes), 10, len=1, replace=FALSE)
  expect_equal(unique(width(gg)), 1)
  
  expect_error(gr.sample(example_genes[1:3], 1e7, len=1, replace=FALSE))

  gg <- gr.sample(example_genes[1:5], c(2,2,3,4,5), len=2, replace=FALSE)
  expect_equal(length(gg), 16)
  expect_equal(sum(width(gg)), 32)
})

test_that("gr.chr", {
  expect_equal(as.character(seqnames(gr.chr(GRanges(c(1,"chrX"), IRanges(c(1,2), 1))))), c("chr1","chrX"))
})

test_that("gr.nochr",{
  expect_identical(gr.nochr(gr.chr(example_genes)), example_genes)
})

test_that("gr.string", {
  expect_that(grepl(":",gr.string(example_genes)[1]), is_true())
  expect_that(grepl("-",gr.string(example_genes)[1]), is_true())
  expect_that(grepl("(+|-)",gr.string(example_genes)[1]), is_true())
})

test_that("gr.fix", {
  gg <- GRanges(c("X",1), IRanges(c(1,2), width=1))
  expect_equal(length(seqlengths(gr.fix(gg, si))), 25)
  expect_equal(length(seqlengths(gr.fix(gg, BSgenome.Hsapiens.UCSC.hg19::Hsapiens))), 95)
})

test_that("gr.flipstrand", {
  expect_identical(as.character(strand(gr.flipstrand(gr))), c("-","+","+"))
})

test_that("gr.flatten", {
  df <- gr.flatten(gr)
  expect_equal(as.character(class(df)), "data.frame")
  expect_identical(colnames(df), c("start", 'end','name'))
  expect_identical(df$start, c(1,4,7))

  df <- gr.flatten(gr, gap=5)
  expect_equal(df$start[2] - df$end[1], 5 + 1)
})

test_that("grlbind", {
  grl.hiC2 <- grl.hiC[1:20]
  mcols(grl.hiC2)$test = 1
  suppressWarnings(gg <- grlbind(grl.hiC2, grl.hiC[1:30]))
  expect_equal(length(gg), 50)
  expect_equal(colnames(mcols(gg)), "test")

  names(grl.hiC) <- NULL
  out <- grlbind(grl.hiC)
  names(out) <- NULL
  expect_identical(out, grl.hiC)

  ## expect error
  expect_error(grlbind('d'))
})

test_that("streduce", {
  gg <- streduce(grl.hiC, pad=10)
  expect_equal(length(gg), length(reduce(gg)))

  gg <- streduce(example_genes, pad=10)
  expect_equal(length(gg), length(reduce(gg)))
})

test_that("grl.pivot", {
  gg <- grl.pivot(grl.hiC)
  expect_equal(as.character(class(gg)), "GRangesList")
  expect_equal(length(gg),2)
  expect_equal(length(gg[[1]]), 10000)
})

test_that("gr.tile", {
  expect_identical(start(gr.tile(gr, w=3)), c(3L, 7L, 13L, 16L))
  expect_equal(length(gr.tile(GRanges())), 0)
})

test_that("grl.string", {
  expect_that(nchar(names(grl.string(grl.hiC[1:5])[1])) > 0, is_true())
  expect_that(grepl(",",grl.string(grl.hiC[1:5])[1]), is_true())
})

test_that("grl.unlist", {
  gg <- grl.unlist(grl.hiC)
  expect_equal(length(gg), length(grl.hiC)*2)
  expect_equal(max(mcols(gg)$grl.iix), 2)
  expect_equal(max(mcols(gg)$grl.ix), length(grl.hiC))

})

test_that("grl.in", {
  gg <- grl.in(grl.hiC[1:100], example_genes)
  expect_equal(length(gg), 100)
})

test_that("gr.fix with null genome", {
  gg <- GRanges(c("X",1), IRanges(c(1,2), width=1))
  es <- structure(c(1L,2L), names=c("X","1"))
  expect_identical(seqlengths(seqinfo(gr.fix(gg))), es)

  es <- structure(c("gen","gen"), names=c("X","1"))
  expect_identical(genome(seqinfo(gr.fix(gg, gname='gen'))), es)

})

test_that("grfo", {
  fo <- example_genes %*% example_dnase
  expect_that(ncol(mcols(fo)) > 2, is_true())
  expect_that(length(fo) > 0, is_true())
})

test_that("gr.simplify", {
  gg <- gr.simplify(gr, pad=4)
  expect_identical(end(gg), c(5L, 16L))
  expect_equal(length(gg), 2)

  gr$field <- c("A","B","B")
  expect_equal(length(gr.simplify(gr, pad=4, field="name")), 3)
  expect_equal(length(gr.simplify(gr, pad=4, field="field")), 2)

  expect_equal(class(gr.simplify(gr, pad=4, field="name", split = TRUE))[1], "GRangesList")
  expect_equal(ncol(mcols((gr.simplify(gr, pad=4, field="name", include.val = FALSE)))), 0)
})

test_that("gr.tile.map", {

  gr1 <- gr.tile(GRanges(1, IRanges(1,100)), w=10)
  gr2 <- gr.tile(GRanges(1, IRanges(1,100)), w=5)
  gg <- gr.tile.map(gr1, gr2, verbose=TRUE)
  expect_equal(length(gg), 10)
  expect_equal(length(unlist(gg)), 20)

})

test_that('ra.overlaps throws error', {
 ss<- split(example_genes, example_genes$name)
 expect_error(ra.overlaps(ss,ss))
})

test_that("ra.overlaps handles empty",{
  ## test empty inputs and no overlaps inputs
  expect_equal(ra.overlaps(GRangesList(), grl1)[1], NA)
  expect_equal(ra.overlaps(grl2[2:3], grl1)[1], NA)
})

test_that("ra.overlaps handles wrong signs", {
  ## make one that overlaps, but wrong signs
  grl3 <- grl1[115]
  strand(grl3[[1]]) <- c("+", "-")
  expect_equal(ra.overlaps(grl3, grl2)[1], NA)
})

test_that('ra.overlaps', {
  grl1 <- grlbind(grl1, grl2)
  ro <- ra.overlaps(grl1, grl2)
  expect_equal(class(ro), "matrix")
  expect_equal(nrow(ro), 252)
  expect_equal(ncol(ro), 2)
  expect_equal(nrow(ra.overlaps(grl2, grl2)), length(grl2))
})


data-raw/prep_UCSC.R
# Output standard human genome seqlengths
#
# Outputs a standard seqlengths for human genome +/- "chr"
# @param hg BSgenome object, perhaps from \code{library(BSgenome.Hsapiens.UCSC.hg19)}
# @param chr Flag for whether to keep "chr". Default FALSE
# @param include.junk Flag for whether to not trim to only 1-22, X, Y, M. Default FALSE
# @return Seqlengths
# @examples
# library(BSgenome.Hsapiens.UCSC.hg19)
# sl <- hg_seqlengths(Hsapiens)
# si <- Seqinfo(seqnames=names(sl), seqlengths=sl, genome="hg19")
# isCircular(si) <- rep(FALSE, length(sl))
# @keywords internal
hg_seqlengths = function(hg, chr = FALSE, include.junk = FALSE)
{
  sl = seqlengths(hg)

  if (!include.junk)
    sl = sl[c(paste('chr', 1:22, sep = ''), 'chrX', 'chrY', 'chrM')]

  if (!chr)
    names(sl) = gsub('chr', '', names(sl))

  return(sl)
}

## prepare a Seqinfo object
library(BSgenome.Hsapiens.UCSC.hg19)
sl <- hg_seqlengths(Hsapiens)
si <- Seqinfo(seqnames=names(sl), seqlengths=sl, genome="hg19")
isCircular(si) <- rep(FALSE, length(sl))

## download data from UCSC Table Browser to include as example data in gUtils

library (rtracklayer)
mySession = browserSession("UCSC")
genome(mySession) <- "hg19"

## DNAaseI hypersensitivity sites
track.name <- "wgEncodeUwDgf"
table.name <- "wgEncodeUwDgfK562Hotspots"
tbl.DNAase <- getTable( ucscTableQuery(mySession, track=track.name, table=table.name))
gr.DNAase <- with(tbl.DNAase, GRanges(gsub("chr(.*?)", "\\1", chrom), IRanges(chromStart, chromEnd), seqinfo=si))
mcols(gr.DNAase) <- tbl.DNAase[,-which(colnames(tbl.DNAase) %in% c("chrom", "chromStart","chromEnd","strand","name","qValue"))]

## RefSeq genes
track.name <- "RefSeq Genes"
table.name <- "refGene"
tbl.genes <- getTable( ucscTableQuery(mySession, track=track.name, table=table.name))
gr.genes <- with(tbl.genes[ix <- !grepl("(hap)|(gl)", tbl.genes$chrom),], GRanges(gsub("chr(.*?)", "\\1", chrom), IRanges(cdsStart, cdsEnd), strand=strand, seqinfo=si))
mcols(gr.genes) <- tbl.genes[ix,-which(colnames(tbl.genes) %in% c("chrom","strand","exonStarts","exonEnds","exonFrames"))]

## import some HiC data
library(LiebermanAidenHiC2009)
library(rtracklayer)
data("HiC_GM_chr14")
chain <- import.chain("/Users/jwala/hg18ToHg19.over.chain")
gr1 <- with(HiC_GM_chr14,GRanges(paste("chr",c(chromosome1,chromosome2),sep=""), IRanges(c(position1, position2),width=1), strand=ifelse(c(strand1, strand2)==0,"+","-"), id=c(seq_along(strand1), seq_along(strand1))))
gr1.hg19 <- gr.fix(gr.nochr(unlist(liftOver(gr1, chain))), si)
tab <- table(gr1.hg19$id)
gr1.hg19 <- gr1.hg19[as.character(gr1.hg19$id) %in% names(tab[tab==2])]
grl.hiC <- GenomicRanges::split(gr1.hg19, gr1.hg19$id)

## make some fake rearrangement data
library(gUtils)
set.seed(137)
grg <- gr.genes
mcols(grg) <- NULL
strand(grg) <- ifelse(runif(length(grg)) > 0.5, "+", "-")
grl <- S4Vectors::split(gr.start(grg), gr.genes$bin)[sapply(split(grg, gr.genes$bin), length)==2]
sam <- sample(length(grl), length(grl), replace=FALSE)
grl1 <- grl[1:(length(grl)/2)]
grl2 <- grl[(length(grl)/2):length(grl)]

## save it
set.seed(137)
gr.genes <- gr.genes[sample(length(gr.genes), 10000)]
gr.DNAase <- gr.DNAase[sample(length(gr.DNAase), 10000)]
grl.hiC <- grl.hiC[sample(length(grl.hiC), 10000)]
save(gr.genes, gr.DNAase, si, grl1, grl2, grl.hiC, file="data/grdata.rda", compress='xz')


rtdocs/figs.R
setwd(file.path(Sys.getenv("GIT_HOME"), "gUtils", "rtdocs"))
HI=150
WI=600

.plot_gr <- function(gro, grn) {

  gro_p <- gro[strand(gro) %in% c("+","*")]
  gro_m <- gro[strand(gro) %in% c("-")]
  grn_p <- grn[strand(grn) %in% c("+","*")]
  grn_m <- grn[strand(grn) %in% c("-")]

  alpha = 0.5
  g <- ggplot()
  if (length(gro_p))
    g <- g + geom_rect(data=as.data.frame(gro_p), aes(xmin=start, xmax=end+1, ymin=4, ymax=4.9), fill='blue', color='black', alpha=alpha)
  if (length(gro_m))
    g <- g + geom_rect(data=as.data.frame(gro_m), aes(xmin=start, xmax=end+1, ymin=3, ymax=3.9), fill='blue', color='black', alpha=alpha)

  if (length(grn_p))
    g <- g + geom_rect(data=as.data.frame(grn_p), aes(xmin=start, xmax=end+1, ymin=1, ymax=1.9), fill='red', color='black', alpha=alpha)
  if (length(grn_m))
    g <- g + geom_rect(data=as.data.frame(grn_m), aes(xmin=start, xmax=end+1, ymin=0, ymax=0.9), fill='red', color='black', alpha=alpha)

  g <- g + theme(axis.text.y=element_blank(), axis.ticks.y=element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_line(colour="gray65", size=0.5, linetype = 'dashed'), axis.ticks.x=element_blank())
  g <- g + scale_x_continuous(breaks=seq(from=1.5, to=20.5, by=1), minor_breaks=seq(from=1,to=20, by=1), labels=1:20) + coord_fixed()
  g <- g + scale_y_continuous(breaks = NULL)
  g <- g + geom_text(data=data.frame(x=c(0,0,0,0), y=c(3.5, 4.5, 0.5, 1.5), text=c("gr(-)", "gr(+)", "(-)","(+)")), aes(x=x,y=y,label=text)) + xlab("") + ylab("")
  g <- g + geom_line(data=data.frame(x=c(1,1), y=c(0,5)), aes(x=x,y=y), color='black')

  .lab_start <- function(grr) { return( (end(grr) - start(grr))/2 + start(grr) + 0.5)}
  ## add labels to boxes
  xs <- c(.lab_start(gro_p), .lab_start(gro_m), .lab_start(grn_p), .lab_start(grn_m))
  ys <- c(rep(4.5, length(gro_p)), rep(3.5, length(gro_m)), rep(1.5, length(grn_p)), rep(0.5, length(grn_m)))
  tx <- c(gro$name, grn$name)

  g <- g + geom_text(data=data.frame(x=xs, y=ys, text=tx), aes(x=x, y=y, label=text))
  return(g)
}

#HI=2
#WI=9
## make the granges
gr  <- GRanges(1, IRanges(c(3,7,13), c(5,9,16)), strand=c('+','-','-'), seqinfo=Seqinfo("1", 25), name=c("A","B","C"))
gr2 <- GRanges(1, IRanges(c(2,4,6), c(3,7,9)), seqinfo=Seqinfo("1", 25), name=c("A","B","C"))
#gr2 <- GRanges(1, IRanges(c(1,9), c(6,14)), seqinfo=Seqinfo("1", 20))
dt <- data.table(seqnames=1, start=c(2,5,10), end=c(3,8,15))

## gr.start
png("rtdocs/figures/gr.start.png", height=HI, width=WI);
print(.plot_gr(gr, gr.start(gr, width=2)) + ggtitle("gr.start(gr, width=2)"));
dev.off()

## gr.start with strand
png("rtdocs/figures/gr.start_wstrand.png", height=HI, width=WI);
print(.plot_gr(gr, gr.start(gr, width=2, ignore.strand=FALSE)) + ggtitle("gr.start(gr, width=2, ignore.strand=FALSE)"));
dev.off()

## gr.end
png("rtdocs/figures/gr.end.png", height=HI, width=WI);
print(.plot_gr(gr, gr.end(gr, width=2)) + ggtitle("gr.end(gr, width=2)"));
dev.off()

## gr.end
png("rtdocs/figures/gr.end_wstrand.png", height=HI, width=WI);
print(.plot_gr(gr, gr.end(gr, width=2, ignore.strand=FALSE)) + ggtitle("gr.end(gr, width=2, ignore.strand=FALSE)"));
dev.off()

## gr.mid
png("rtdocs/figures/gr.mid.png", height=HI, width=WI);
print(.plot_gr(gr, gr.mid(gr)) + ggtitle("gr.mid(gr)"));
dev.off()

## shift
png("rtdocs/figures/shift.png", height=HI, width=WI);
print(.plot_gr(gr, GenomicRanges::shift(gr,2)) + ggtitle("GenomicRanges::shift(gr,2)"));
dev.off()

## flank
png("rtdocs/figures/flank.png", height=HI, width=WI);
print(.plot_gr(gr, GenomicRanges::flank(gr,width=1)) + ggtitle("GenomicRanges::flank(gr,width=1)"));
dev.off()

## flank w/start
png("rtdocs/figures/flank_start.png", height=HI, width=WI);
print(.plot_gr(gr, GenomicRanges::flank(gr,start=FALSE, width=1)) + ggtitle("GenomicRanges::flank(gr,start=FALSE, width=1)"));
dev.off()

## streduce
png("rtdocs/figures/streduce.png", height=HI, width=WI)
grn = streduce(gr+2)
grn$name <- c("ABC")
print(.plot_gr(gr+2, grn) + ggtitle("streduce(gr+2)"));
dev.off()

## gr.flatten
png("rtdocs/figures/gr.flatten.png", height=HI, width=WI)
grn <- with(gr.flatten(gr), GRanges(1, IRanges(start, end), name=name))
print(.plot_gr(gr, grn) + ggtitle("gr.flatten(gr)"));
dev.off()

## gr.simplify
png("rtdocs/figures/gr.simplify.png", height=HI, width=WI)
grn <- gr.simplify(gr2)
grn$name = "ABC"
print(.plot_gr(gr2, grn) + ggtitle("gr.simplify(gr2)"));
dev.off()

## gr.flipstrand
png("rtdocs/figures/gr.flipstrand.png", height=HI, width=WI)
print(.plot_gr(gr, gr.flipstrand(gr)) + ggtitle("gr.flipstrand(gr)"));
dev.off()

## gr.tile
png("rtdocs/figures/gr.tile.png", height=HI, width=WI)
grn=gr.tile(gr,w=2)
grn$name = c("A1","A2","B1","B2","C1","C2")
print(.plot_gr(gr, grn) + ggtitle("gr.tile(gr,w=2)"));
dev.off()

## gr.sample
set.seed(137)
png("rtdocs/figures/gr.sample.png", height=HI, width=WI)
print(.plot_gr(gr, gr.sample(gr, 3, len=2, replace=TRUE)) + ggtitle("streduce(gr+2)"));
dev.off()

g <- ggplot() +
  geom_rect(data=as.data.frame(gr), aes(xmin=start, xmax=end, ymin=1, ymax=2), fill='blue') +
  geom_rect(data=as.data.frame(gr.sample(gr, 3, len=2, replace=TRUE)), aes(xmin=start, xmax=end, ymin=0, ymax=0.95), fill='red') +
  theme_bw() + theme(axis.text.y=element_blank(), axis.ticks.y=element_blank()) + scale_x_continuous(breaks=0:20, labels=0:20) + coord_fixed()
pdf("figures/gr.sample.pdf", height=HI, width=WI); print(g); dev.off()
system.call("convert figures/gr.sample.pdf figures/gr.sample.png")

## gr.rand
set.seed(137)
g <- ggplot() +
  geom_rect(data=as.data.frame(gr), aes(xmin=start, xmax=end, ymin=1, ymax=2), fill='blue') +
  geom_rect(data=as.data.frame(gr.rand(c(2,5,3), seqinfo(gr))), aes(xmin=start, xmax=end, ymin=0, ymax=0.95), fill='red') +
  theme_bw() + theme(axis.text.y=element_blank(), axis.ticks.y=element_blank()) + scale_x_continuous(breaks=0:20, labels=0:20) + coord_fixed()
pdf("figures/gr.rand.pdf", height=HI, width=WI); print(g); dev.off()
system.call("convert figures/gr.rand.pdf figures/gr.rand.png")

## gr.simplify
gr3 <- GRanges(1, IRanges(c(1,4,8), c(4,8,10)))
df <- data.frame(start=c(1,4,8), end=c(3.95, 7.96, 10))
g <- ggplot() +
  geom_rect(data=df, aes(xmin=start, xmax=end, ymin=1, ymax=2), fill='blue') +
  geom_rect(data=as.data.frame(gr.simplify(gr3)), aes(xmin=start, xmax=end, ymin=0, ymax=0.95), fill='red') +
  theme_bw() + theme(axis.text.y=element_blank(), axis.ticks.y=element_blank()) + scale_x_continuous(breaks=0:20, labels=0:20) + coord_fixed()
pdf("figures/gr.simplify.pdf", height=HI, width=WI); print(g); dev.off()
system.call("convert figures/gr.simplify.pdf figures/gr.simplify.png")

## gr.tile
df1 <- data.frame(xmin=c(0,3,6), xmax=c(4,7,10), ymin=c(0,1,2), ymax=c(1,2,3))
df2 <- data.frame(xmin=1, xmax=9, ymin=3, ymax=4)
g <- ggplot() +
  geom_rect(data=df1, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax), fill='blue') +
  geom_rect(data=df2, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax), fill='red') +
  theme_bw() + theme(axis.text.y=element_blank(), axis.ticks.y=element_blank()) + scale_x_continuous(breaks=0:20, labels=0:20) + coord_fixed()
pdf("figures/gr.tile.pdf", height=HI, width=WI); print(g); dev.off()
system.call("convert figures/gr.tile.pdf figures/gr.tile.png")

## gr.refactor
gr3 <- GRanges(1, IRanges(c(4,11), c(6, 20)))
g <- ggplot() + geom_rect(data=as.data.frame(gr3), aes(xmin=start, xmax=end, ymin=1, ymax=2), fill='blue') + geom_rect(data=as.data.frame(gr.refactor(gr3, "3")), aes(xmin=start, xmax=end, ymin=0, ymax=0.9), fill='red') + theme_bw() + theme(axis.text.y=element_blank(), axis.ticks.y=element_blank()) + scale_x_continuous(breaks=0:20, labels=0:20) + coord_fixed()
pdf("figures/gr.refactor.pdf", height=HI, width=WI); print(g); dev.off()
system.call("convert figures/gr.refactor.pdf figures/gr.refactor.png")

## flank start=FALSE
g <- ggplot() + geom_rect(data=as.data.frame(gr), aes(xmin=start, xmax=end, ymin=1, ymax=2), fill='blue') + geom_rect(data=as.data.frame(flank(gr,width=2, start=FALSE)), aes(xmin=start, xmax=end, ymin=0, ymax=0.9), fill='red') + theme_bw() + theme(axis.text.y=element_blank(), axis.ticks.y=element_blank()) + scale_x_continuous(breaks=0:20, labels=0:20) + coord_fixed()

## gr.round
g <- ggplot() + geom_rect(data=as.data.frame(gr), aes(xmin=start, xmax=end, ymin=1, ymax=2), fill='blue') + geom_rect(data=as.data.frame(gr.round(gr, gr)), aes(xmin=start, xmax=end, ymin=0, ymax=0.9), fill='red') + theme_bw() + theme(axis.text.y=element_blank(), axis.ticks.y=element_blank()) + scale_x_continuous(breaks=0:20, labels=0:20) + coord_fixed()


